cpp/src/util/file_io.hpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <raft/core/error.hpp>
+
+#include <algorithm>
+#include <cstring>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include <errno.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+namespace cuvs::util {
+
+/**
+ * @brief RAII wrapper for POSIX file descriptors
+ *
+ * Manages file descriptor lifecycle with automatic cleanup.
+ * Non-copyable, move-only.
+ */
+class file_descriptor {
+ public:
+  explicit file_descriptor(int fd = -1) : fd_(fd) {}
+
+  file_descriptor(const std::string& path, int flags, mode_t mode = 0644)
+    : fd_(open(path.c_str(), flags, mode)), path_(path)
+  {
+    if (fd_ == -1) {
+      RAFT_FAIL("Failed to open file: %s (errno: %d, %s)", path.c_str(), errno, strerror(errno));
+    }
+  }
+
+  file_descriptor(const file_descriptor&)            = delete;
+  file_descriptor& operator=(const file_descriptor&) = delete;
+
+  file_descriptor(file_descriptor&& other) noexcept
+    : fd_{std::exchange(other.fd_, -1)}, path_{std::move(other.path_)}
+  {
+  }
+
+  file_descriptor& operator=(file_descriptor&& other) noexcept
+  {
+    std::swap(this->fd_, other.fd_);
+    std::swap(this->path_, other.path_);
+    return *this;
+  }
+
+  ~file_descriptor() noexcept { close(); }
+
+  [[nodiscard]] int get() const noexcept { return fd_; }
+  [[nodiscard]] bool is_valid() const noexcept { return fd_ != -1; }
+
+  void close() noexcept
+  {
+    if (fd_ != -1) {
+      ::close(fd_);
+      fd_ = -1;
+    }
+  }
+
+  [[nodiscard]] int release() noexcept
+  {
+    const int fd = fd_;
+    fd_          = -1;
+    return fd;
+  }
+
+  [[nodiscard]] std::string get_path() const { return path_; }
+
```

Comment:
A helper function to return a C++ stream.
```suggestion
  /**
   * @brief Create an input stream from this file descriptor
   *
   * Creates an istream that reads directly from the file descriptor using POSIX read().
   * The original descriptor remains valid and unchanged (we duplicate it internally).
   * Returns the stream by value (uses move semantics).
   *
   * @return fd_istream (movable istream)
   */
  [[nodiscard]] fd_istream make_istream() const
  {
    RAFT_EXPECTS(is_valid(), "Invalid file descriptor");

    // Duplicate the fd to avoid consuming the original
    int dup_fd = dup(fd_);
    RAFT_EXPECTS(dup_fd != -1, "Failed to duplicate file descriptor");

    // Create stream that owns the duplicated fd
    // Returned by value, uses move semantics
    return fd_istream(dup_fd);
  }
```